src/utils/dateTimeUtils.ts

```diff
@@ -0,0 +1,74 @@
+import { DataTypes } from "../types";
+import { dateTimeRegexMap, daysOfWeek, defaultDate, invalidDateTimeErr, longDateFormat, longTimeFormat, milliSecPerDay, months, monthsbeforeLeap, numberOfDaysTillExcelBeginYear,  shortTimeFormat } from "./constants";
+
+
+const convertToExcelDate = (dateTime: string, dataType: DataTypes) => {
+    let dateTimeStr: string = dateTime;
+    if (dataType == DataTypes.shortTime || dataType == DataTypes.longTime) {
+        // Excel assumes that the time is in the 31/12/1899 if none is specified
+        dateTimeStr = defaultDate + dateTimeStr;
+    }
+
+    const localDate: Date = new Date(dateTimeStr);
+    if (isNaN(localDate.getTime())) {
+        throw new Error(invalidDateTimeErr);
+    }
+
+    const globalDate = new Date(Date.UTC(localDate.getFullYear(), localDate.getMonth(), 
+        localDate.getDate(), localDate.getHours(), localDate.getMinutes(), localDate.getSeconds())).getTime();
+    // Excel incorrectly assumes that the year 1900 is a leap year. This is a workaround for that
+    if ((localDate.getFullYear() == 1900 && localDate.getMonth() <= monthsbeforeLeap) || localDate.getFullYear() < 1900) {
+        return ((globalDate + numberOfDaysTillExcelBeginYear*milliSecPerDay) / (milliSecPerDay)) - 1;
+    }
+
+    return (globalDate + numberOfDaysTillExcelBeginYear*milliSecPerDay) / (milliSecPerDay);
+};
+
+const getFormatCode = (format: DataTypes): string => {
+    switch (format) {
+        case DataTypes.longDate:
+            return longDateFormat;
+        case DataTypes.longTime:
+            return longTimeFormat;
+        case DataTypes.shortTime:
+            return shortTimeFormat;
+        default:
+           throw new Error (invalidDateTimeErr);
+    }
+};
+
+export const detectDateTimeFormat = (dateTime: string): DataTypes|undefined => { 
+    for (const [regex, dataType] of dateTimeRegexMap) {
+        if (regex.test(dateTime)) {
+            return validateDate(dateTime, dataType);
```

Comment:
The regex checks the date format, but doesn't check if the date is valid. For ex - 4/31 or wrong day of the week (Sunday Sept 20, 2023 instead of Wednesday) etc.